Add initial GetDeclaredSymbol support for inferred VariableDeclaratorSyntax and SingleVariableDesignationSyntax nodes

333fred opened the pull request:
Add GetDeclaredSymbol support for nullability in variable declarations. Not handled are linq or foreach iteration variable declarations. @dotnet/roslyn-compiler for review. Foreach and linq are tracked by https://github.com/dotnet/roslyn/issues/37321.
/cc @jasonmalinowski @ryzngard.
Fixes https://github.com/dotnet/roslyn/issues/32459.
Fixes https://github.com/dotnet/roslyn/issues/30150.

333fred commented:
@dotnet/roslyn-compiler for review please.

cston commented on src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs:
```diff
@@ -9,6 +9,26 @@ namespace Microsoft.CodeAnalysis.CSharp
+            internal readonly NullableWalker.SnapshotManager? SnapshotManager;
+            internal readonly ImmutableDictionary<Symbol, TypeWithAnnotations>? AnalyzedVariableDeclarations;
+
+            internal InitialState(CSharpSyntaxNode syntax, BoundNode? bodyOpt = null, ExecutableCodeBinder? binder = null, NullableWalker.SnapshotManager? snapshotManager = null, ImmutableDictionary<Symbol, TypeWithAnnotations>? analyzedVariableDeclarations = null)
```
>NullableWalker.SnapshotManager? snapshotManager = null, ImmutableDictionary<Symbol, TypeWithAnnotations>? analyzedVariableDeclarations = null [](start = 123, length = 141)

Are these arguments used?

333fred replied:
Yes, they're used in MethodCompiler.

---
In reply to: [305047427](https://github.com/dotnet/roslyn/pull/37322#discussion_r305047427) [](ancestors = 305047427)

cston commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1049,25 +1049,22 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     if (body != null)
                     {
-                        (SyntaxNode Syntax, BoundNode Body, ExecutableCodeBinder Binder, NullableWalker.SnapshotManager SnapshotManager) forSemanticModelToUseInLambda = forSemanticModel;
+                        MethodBodySemanticModel.InitialState forSemanticModelToUseInLambda = forSemanticModel;
```
>forSemanticModelToUseInLambda [](start = 61, length = 29)

Is this local needed, or could we use `forSemanticModel` directly?

333fred replied:
I _think_ we could use `forSemanticModel` directly, but I didn't want to change it just in case this was doing something I didn't understand with capturing.

---
In reply to: [305051587](https://github.com/dotnet/roslyn/pull/37322#discussion_r305051587) [](ancestors = 305051587)

jcouv commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1921,22 +1935,37 @@ void ensureSpeculativeNodeBound()
         }
 
-        protected abstract BoundNode RewriteNullableBoundNodesWithSnapshots(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, bool takeSnapshots, out NullableWalker.SnapshotManager snapshotManager);
+        protected abstract BoundNode RewriteNullableBoundNodesWithSnapshots(BoundNode boundRoot, Binder binder, DiagnosticBag diagnostics, bool takeSnapshots, out NullableWalker.SnapshotManager snapshotManager, out ImmutableDictionary<Symbol, TypeWithAnnotations> analyzedVariableTypes);
```
>RewriteNullableBoundNodesWithSnapshots [](start = 37, length = 38)

nit: xml doc on central API would be useful

jcouv commented on src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs:
```diff
@@ -9,6 +9,26 @@ namespace Microsoft.CodeAnalysis.CSharp
     internal sealed class MethodBodySemanticModel : MemberSemanticModel
     {
+#nullable enable
+        internal readonly struct InitialState
```
>InitialState [](start = 33, length = 12)

nit: xml doc would be nice here too. What is this for?

jcouv commented on src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs:
```diff
@@ -28,15 +48,14 @@ internal sealed class MethodBodySemanticModel : MemberSemanticModel
+            if (initialState.Body != null)
             {
-                result.UnguardedAddBoundTreeForStandaloneSyntax(syntax, boundNode, snapshotManager);
+                result.UnguardedAddBoundTreeForStandaloneSyntax(initialState.Syntax, initialState.Body, initialState.SnapshotManager, initialState.AnalyzedVariableDeclarations is { } variables ? GetUpdatedLocalSymbols(variables) : null);
```
>{ } [](start = 179, length = 3)

`object` preferred over `{ }`

333fred replied:
I can't. `{ }` preserves the existing type, `object` does not.

---
In reply to: [305055503](https://github.com/dotnet/roslyn/pull/37322#discussion_r305055503) [](ancestors = 305055503)

jcouv commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1049,25 +1049,22 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                             var semanticModel = (SyntaxTreeSemanticModel)_compilation.GetSemanticModel(syntax.SyntaxTree);
 
-                            if (forSemanticModelToUseInLambda.Syntax != null)
+                            if (forSemanticModelToUseInLambda.Syntax is { } semanticModelSyntax)
```
>{ } [](start = 72, length = 3)

`object`

333fred replied:
I can't, for the same reason as the other one.

---
In reply to: [305055943](https://github.com/dotnet/roslyn/pull/37322#discussion_r305055943) [](ancestors = 305055943)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -450,7 +450,8 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
                 snapshotBuilderOpt: null,
-                returnTypesOpt: null);
+                returnTypesOpt: null,
+                out _);
```
>out [](start = 16, length = 3)

nit: consider naming this argument

333fred replied:
No longer relevant.

---
In reply to: [305056178](https://github.com/dotnet/roslyn/pull/37322#discussion_r305056178) [](ancestors = 305056178)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -550,7 +554,8 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
                 snapshotBuilderOpt: null,
-                returnTypesOpt: null);
+                returnTypesOpt: null,
+                out _);
```
>out _); [](start = 16, length = 7)

nit: ditto

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -1198,5 +1208,406 @@ static void verifySymbolInfo(IMethodSymbol methodSymbol, PublicNullableAnnotatio
+        }
+
+        [Fact]
+        public void GetDeclaredSymbol_Using()
```
>GetDeclaredSymbol_Using [](start = 20, length = 23)

consider adding an `await using` variant in test

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -1198,5 +1208,406 @@ static void verifySymbolInfo(IMethodSymbol methodSymbol, PublicNullableAnnotatio
+        }
+
+        [Fact]
+        public void GetDeclaredSymbol_Switch()
```
>GetDeclaredSymbol_Switch [](start = 20, length = 24)

consider adding a switch expression variant in test

333fred replied:
There is one :)

---
In reply to: [305060750](https://github.com/dotnet/roslyn/pull/37322#discussion_r305060750) [](ancestors = 305060750)

jcouv commented:
>         public void SpeakableInference_VarInference()

Can this test be completed and the comment removed?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:4870 in 43e7375. [](commit_id = 43e73750efd10b82182a65d408885f9e3177a228, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -36267,8 +36265,8 @@ static void F(string? s)
             var declarator = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
             var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarator);
             // https://github.com/dotnet/roslyn/issues/26198 public API should show inferred nullability
```
>// https://github.com/dotnet/roslyn/issues/26198 public API should show inferred nullability [](start = 12, length = 92)

is this comment still relevant?

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -44531,11 +44518,11 @@ public void Default_TClass()
             var declarators = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().ToArray();
             var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarators[0]);
             // https://github.com/dotnet/roslyn/issues/26198 public API should show inferred nullability
```
>// https://github.com/dotnet/roslyn/issues/26198 public API should show inferred nullability [](start = 12, length = 92)

is this comment still relevant?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -610,15 +617,16 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             VariableState initialState,
-            SnapshotManager.Builder snapshotBuilderOpt)
+            SnapshotManager.Builder snapshotBuilderOpt,
+            out ImmutableDictionary<Symbol, TypeWithAnnotations> variableTypes)
```
>out ImmutableDictionary<Symbol, TypeWithAnnotations> variableTypes [](start = 12, length = 66)

Consider having the caller pass in an optional mutable `Dictionary<Symbol, TypeWithAnnotations>` so we're not allocating for callers that ignore this value.

cston commented:
>             }

The caller allocated the `NullableWalker` and should be responsible for calling `Free()`.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:650 in 43e7375. [](commit_id = 43e73750efd10b82182a65d408885f9e3177a228, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -628,11 +636,13 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
                 ImmutableArray<PendingBranch> returns = walker.Analyze(ref badRegion, initialLocalState);
                 snapshotBuilderOpt?.ExitWalker(walker.SaveSharedState(), previousSlot);
                 diagnostics?.AddRange(walker.Diagnostics);
+                variableTypes = walker._variableTypes.ToImmutableDictionary();
```
>variableTypes = walker._variableTypes.ToImmutableDictionary(); [](start = 16, length = 62)

Should be moved to the caller so the caller can avoid the allocation if not needed.

333fred replied:
This can't be done. Freeing the walker frees the variable types PooledDictionary, and I don't think it's correct to move responsibility for freeing this dictionary out of the main walker itself.

---
In reply to: [305068984](https://github.com/dotnet/roslyn/pull/37322#discussion_r305068984) [](ancestors = 305068984)

cston replied:
This is a `private` method. The only callers are in the walker.

---
In reply to: [305452556](https://github.com/dotnet/roslyn/pull/37322#discussion_r305452556) [](ancestors = 305452556,305068984)

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -397,6 +397,23 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
+        /// type should be identical to the original except for nullability.
+        /// </summary>
+        internal LocalSymbol WithAnalyzedType(TypeWithAnnotations analyzedType)
+        {
```
>{ [](start = 8, length = 1)

If the current type is `analyzedType`, we should return the original instance.

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -701,26 +699,30 @@ class C
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
 
             comp.VerifyDiagnostics();
             comp.VerifyAnalyzerDiagnostics(new[] { new NullabilityPrinter() }, null, null, true,
```
>VerifyAnalyzerDiagnostics [](start = 17, length = 25)

Does `VerifyAnalyzerDiagnostics` generate the actual result with comments? In short, can we add the diagnostic messages here easily?

333fred replied:
It does not, this is the result it generates.

---
In reply to: [305071650](https://github.com/dotnet/roslyn/pull/37322#discussion_r305071650) [](ancestors = 305071650)

333fred commented:
@jcouv @cston addressed feedback. I rewrote this to be much simpler than was originally implemented, using existing saved state instead of adding new state.

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -26,6 +27,7 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
         // The bound nodes associated with a syntax node, from highest in the tree to lowest.
         private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedNodeMap = new Dictionary<SyntaxNode, ImmutableArray<BoundNode>>();
         private Dictionary<SyntaxNode, BoundStatement> _lazyGuardedSynthesizedStatementsMap;
+        private readonly ConcurrentDictionary<LocalSymbol, LocalSymbol> _analyzedVariableTypesOpt = new ConcurrentDictionary<LocalSymbol, LocalSymbol>();
```
>new ConcurrentDictionary<LocalSymbol, LocalSymbol>(); [](start = 100, length = 53)

Should this be allocated only if `NullableSemanticAnalysisEnabled`?

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -655,7 +657,33 @@ private LocalSymbol GetDeclaredLocal(CSharpSyntaxNode declarationSyntax, SyntaxT
+                        LocalSymbol adjustedLocal;
+                        if (Compilation.NullableSemanticAnalysisEnabled)
+                        {
+                            if (!_analyzedVariableTypesOpt.TryGetValue(local, out adjustedLocal))
```
>if (!_analyzedVariableTypesOpt.TryGetValue(local, out adjustedLocal)) [](start = 28, length = 69)

Please use: `_analyzedVariableTypesOpt.GetOrAdd(local, (local, arg) => { ... }, ...);`

333fred replied:
I can't. This project targets netstandard2.0, that api was added to netstandard until 2.1.

---
In reply to: [305589670](https://github.com/dotnet/roslyn/pull/37322#discussion_r305589670) [](ancestors = 305589670)

cston replied:
An instance field could replace the lambda.
```
return _analyzedVariableTypesOpt.GetOrAdd(local, _getAnalyzedLocal);
```

---
In reply to: [306051195](https://github.com/dotnet/roslyn/pull/37322#discussion_r306051195) [](ancestors = 306051195,305589670)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -84,6 +73,29 @@ private SnapshotManager(ImmutableArray<SharedWalkerState> walkerSharedStates, Im
+            internal ImmutableDictionary<Symbol, TypeWithAnnotations> GetVariableTypesForPosition(int position)
+            {
+                var snapshot = GetSnapshotForPosition(position);
+                var globalState = _walkerSharedStates[snapshot.SharedStateIndex];
```
>globalState [](start = 20, length = 11)

"sharedState"?

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -632,7 +630,10 @@ public class C
             void verifyCompilation(CSharpCompilation compilation)
             {
                 var members = memberFunc(compilation);
-                AssertEx.Equal(expectedNullabilities, members.Select(nullabilityFunc));
```
>AssertEx.Equal(expectedNullabilities, members.Select(nullabilityFunc)); [](start = 16, length = 71)

Why change? The loop will only report the first failure rather than all.

cston replied:
It looks like there is a `FormatException` in `GetAssertMessage<T>()`. We should use `typeof(T) == typeof(byte)` there rather than `expected is IEnumerable<byte>`.

---
In reply to: [305590024](https://github.com/dotnet/roslyn/pull/37322#discussion_r305590024) [](ancestors = 305590024)

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -71,6 +73,11 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
+
+            if (Compilation.NullableSemanticAnalysisEnabled)
+            {
+                _analyzedVariableTypesOpt = new ConcurrentDictionary<LocalSymbol, LocalSymbol>();
```
>new ConcurrentDictionary [](start = 44, length = 24)

Consider delaying allocation until needed in `GetDeclaredLocal()`.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -397,6 +397,28 @@ internal void SetTypeWithAnnotations(TypeWithAnnotations newType)
+        /// </summary>
+        internal LocalSymbol WithAnalyzedType(TypeWithAnnotations analyzedType)
+        {
+            Debug.Assert(analyzedType.Equals(TypeWithAnnotations, TypeCompareKind.AllNullableIgnoreOptions | TypeCompareKind.IgnoreTupleNames));
```
>TypeCompareKind.IgnoreTupleNames [](start = 109, length = 32)

Are there cases where the analyzed type differs from the original by tuple names? Is that a bug?

333fred replied:
Yes, tracked by [https://github.com/dotnet/roslyn/issues/34993](https://github.com/dotnet/roslyn/issues/34993).

---
In reply to: [306385539](https://github.com/dotnet/roslyn/pull/37322#discussion_r306385539) [](ancestors = 306385539)

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -666,6 +661,7 @@ private LocalSymbol GetDeclaredLocal(CSharpSyntaxNode declarationSyntax, SyntaxT
                         LocalSymbol adjustedLocal;
                         if (Compilation.NullableSemanticAnalysisEnabled)
                         {
+                            _analyzedVariableTypesOpt ??= new ConcurrentDictionary<LocalSymbol, LocalSymbol>();
```
Probably should use `Interlocked.CompareExchange()` to avoid populating multiple dictionaries.

333fred replied:
I'm going to not lazily initialize this. @agocke mentioned that our lock-free code policy at this point is to get the entire team to review. If the concurrent dictionary shows up in a trace, we can lazily initialize at that point.

---
In reply to: [306527725](https://github.com/dotnet/roslyn/pull/37322#discussion_r306527725) [](ancestors = 306527725)

333fred replied:
If you feel strongly about this, I can use a `Lazy<T>` on the concurrent dictionary, but that feels like overkill imo.

---
In reply to: [306933090](https://github.com/dotnet/roslyn/pull/37322#discussion_r306933090) [](ancestors = 306933090,306527725)

agocke replied:
Clarification: I didn't actually look at this code. The policy is we have to review new lock-free data structures together -- all existing patterns are fine. In this case, this would be equivalent to a lazy initialization, which is a well-established pattern.

The point of the policy wasn't to prevent us from using our standard (well-reviewed) lazy initialization patterns, just to prevent us from adding clever new patterns that have very subtle threading requirements.

333fred commented:
@cston @jcouv addressed feedback.

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -511,6 +511,7 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
             var newSnapshotBuilder = takeNewSnapshots ? new SnapshotManager.Builder() : null;
             var (walker, initialState, symbol) = originalSnapshots.RestoreWalkerToAnalyzeNewNode(position, node, binder, analyzedNullabilities, updatedMethodSymbols, newSnapshotBuilder);
             Analyze(walker, symbol, diagnostics: null, initialState, snapshotBuilderOpt: newSnapshotBuilder);
+            walker.Free();
```
>walker.Free(); [](start = 12, length = 14)

try/finally?

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs:
```diff
@@ -1198,5 +1208,454 @@ static void verifySymbolInfo(IMethodSymbol methodSymbol, PublicNullableAnnotatio
+        }
+
+        [Fact]
+        public void GetDeclaredSymbol_OutVariable()
```
>GetDeclaredSymbol_OutVariable [](start = 20, length = 29)

A variant of this test with type inference would be great:
`void Out<T>(T input, out T output)`

jcouv commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -44413,11 +44407,11 @@ static void Main()
+            Assert.Equal(NullableAnnotation.Annotated, symbol.TypeWithAnnotations.NullableAnnotation);
             symbol = (LocalSymbol)model.GetDeclaredSymbol(declarators[1]);
             Assert.Equal("System.Int32", symbol.TypeWithAnnotations.ToTestDisplayString(true));
+            // https://github.com/dotnet/roslyn/issues/34412 tracks standardizing the representation of value types
```
>// https://github.com/dotnet/roslyn/issues/34412 tracks standardizing the representation of value types [](start = 12, length = 103)

When Chuck and you merge your PRs, this will be a conflict. The second person to commit should make sure to re-run CI and remove the comment

jcouv commented on src/Compilers/CSharp/Portable/Compilation/MethodBodySemanticModel.cs:
```diff
@@ -9,6 +8,27 @@ namespace Microsoft.CodeAnalysis.CSharp
     {
+#nullable enable
+        /// <summary>
+        /// Initial state for a MethodBodySemanticModel. Shared between here and the <see cref="MethodCompiler"/>.
```
>Initial state for a MethodBodySemanticModel. Shared between here and the <see cref="MethodCompiler"/>. [](start = 12, length = 102)

nit: Consider "Used to make a MethodBodySemanticModel and we already have some/all the parts we'll need" or something along those lines

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs:
```diff
@@ -84,6 +73,29 @@ private SnapshotManager(ImmutableArray<SharedWalkerState> walkerSharedStates, Im
+
+                if (snapshotIndex < 0)
+                {
+                    // BinarySearch returns the next higher position. Always take the one most close but behind the requested position
```
>most close [](start = 90, length = 10)

nit: closest

gafter commented:
Too late for 16.3.  Please change milestone.

333fred commented:
No, this is specifically being taken for 16.3.

---
In reply to: [514443364](https://github.com/dotnet/roslyn/pull/37322#issuecomment-514443364) [](ancestors = 514443364)

cston commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -71,6 +73,7 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
             _speculatedPosition = speculatedPosition;
 
             _operationFactory = new Lazy<CSharpOperationFactory>(() => new CSharpOperationFactory(this));
+            _analyzedVariableTypesOpt = new ConcurrentDictionary<LocalSymbol, LocalSymbol>();
```
>new ConcurrentDictionary [](start = 40, length = 24)

We should only allocate if `Compilation.NullableSemanticAnalysisEnabled`, as in iteration 3.

jcouv commented:
@333fred There are conflicts

jcouv commented:
@333fred Go ahead and merge once CI is green.
